app/models/activities/fetcher.rb

```diff
@@ -114,6 +115,16 @@ def eager_load_associations(events)
       end
     end
 
+    def filter_by_visibility(events)
+      events.reject! do |event|
+        if OpenProject::FeatureDecisions.comments_with_restricted_visibility_active?
+          event.journal.restricted? && !@user.allowed_in_project?(:view_comments_with_restricted_visibility, event.project)
+        else
+          event.journal.restricted?
+        end
+      end
```

Comment:
This here contradicts what is written in the `restricted_visible` scope where the check is on the work package level as well. 

Additionally, it will be slow without benefits I can see. Adapting the query for fetching the journals in the first place is the way to go, I think.

To this end, a subquery can be added when fetching the journals. That could either be

```
WorkPackage.allowed_to(User.current, :view_comments_with_restricted_visibility)
```

or

```
Project.allowed_to(User.current, :view_comments_with_restricted_visibility)
```

depending on whether the permission can or cannot be assigned to work package roles as well.

Reply:
@akabiru fyi, I'll look into this one, next

Reply:
The fix https://github.com/opf/openproject/pull/18367